ubirch/ubirch_ks.py

```diff
@@ -88,46 +138,202 @@ def insert_ed25519_keypair(self, uuid: UUID, vk: VerifyingKey, sk: SigningKey) -
+
+    def find_signing_key(self, uuid: UUID) -> ed25519.SigningKey or ecdsa.SigningKey:
+        """!
+        Find the signing key from the keystore for this UUID.
+        @param uuid The UUID of the device
+        @return The ed25519 / ecdsa signing key if found. Else returns `None`
+        """
+        # try to find a matching sk for the uuid
+        try:
+            sk : PrivateKeyEntry = self._ks.private_keys['pke_' + uuid.hex]
+        except KeyError as e:
+            # there is no sk for the given uuid
             return None
+        
+        # check whether the entry is encrypted
+        if sk.is_decrypted() == False:
+            sk.decrypt(self._ks_password)
+
+        # check the _OID to identify the key type
+        if sk._algorithm_oid == EDDSA_OID:
+            return ed25519.SigningKey(sk.pkey)
+        elif sk._algorithm_oid == ECDSA_OID:
+            # ==================================== IMPORTANT ====================================
+            #   The used curve as well as the used hash function have to be explicitly set here
+            #   to match the ones used in create_ecdsa_keypair(), otherwise the ._from_string()
+            #   function will throw exceptions because of unexpected/wrong keystr lengths (...)
+            # ===================================================================================
+            return ecdsa.SigningKey.from_string(sk.pkey, curve=ecdsa.NIST256p, hashfunc=hashlib.sha256)
+        else:
+            raise Exception("stored key with unknown algorithm OID: '{}'".format(sk._algorithm_oid))
+
+    def _find_cert(self, uuid: UUID) -> ECDSACertificate or ED25519Certificate:
+        """!
+        Find the stored cert for uuid
+        @param uuid The UUID of the device
+        @return The stored ECDSACertificate or ED25519Certificate if found. Else returns `None`
+        """
+        cert = None
+
+        if self.exists_verifying_key(uuid) == True:
+            # try to get the edd key first
+            try:
+                cert = ED25519Certificate(
+                    self._ks.certs[uuid.hex].alias,
+
+                    # the ED25519Certificate requires an ed25519.VerifyingKey
+                    ed25519.VerifyingKey(self._ks.certs[uuid.hex].cert)
+                )
+            except KeyError:
+                pass
+
+            # no edd key found, try to get ecd
+            try:
+                cert = ECDSACertificate(
+                    self._ks.certs[uuid.hex + '_ecd'].alias,
 
-        cert = self._ks.certs[uuid.hex]
-        vk = VerifyingKey(cert.cert)
+                    # the ECDSACertifcate requires an ecdsa.VerifyingKey
+                    ecdsa.VerifyingKey.from_string(
+                        self._ks.certs[uuid.hex + '_ecd'].cert,
+                        hashfunc=hashlib.sha256, curve=ecdsa.NIST256p
+                    )
+                )
+            except KeyError:
+                pass
+
+        return cert
```

Comment:
This could be simplified to
```suggestion
        cert = self._ks.certs.get(uuid.hex)

        if cert is not None:
            return cert

        return self._ks.certs.get(uuid.hex + '_ecd')
```

If we can drop the `'_ecd'` suffix, it could even be a one-liner
```suggestion
       return self._ks.certs.get(uuid.hex)
```

Reply:
thanks I simplified the code and wrote a test for that

Reply:
turns out, this does not work for ed25519 and ecdsa keys at the same time. When loading the key, the type of key will be recognized by your method. This later leads to errors, since ed25519 has method `.to_bytes()` whicle ecdsa has method `.to_string()` and this is only one of the errors

Reply:
with your changes the `find_veriffying_key()` method returns `NoneType`